actionpack/lib/action_dispatch/middleware/exception_wrapper.rb

```diff
@@ -168,19 +173,56 @@ def clean_backtrace(*args)
         end
       end
 
+      def extract_source(trace)
+        # ErrorHighlight.spot with backtrace_location keyword is available since error_highlight 0.4.0
+        if defined?(ErrorHighlight) && Gem::Version.new(ErrorHighlight::VERSION) >= Gem::Version.new("0.4.0")
```

Comment:
So I'm not too concerned by performance for what is essentially a dev only error reporting, but for clarity it may make sense to extract that version check in a class method and memoize it. So that you can use the same logic here and in the template.

Reply:
Or in a constant, as this could be checked at load time.

Reply:
Actually, they are not completely the same. The tip message should be displayed only when `ErrorHighlight` *is* available but its version is lower than 0.4.0. Otherwise, it will be shown on Ruby 2.7 or TruffleRuby, which are not supported by error_highlight.

I created a patch, but it is larger than I expected because it requires passing a new argument to ActionView of the error page. I'm not sure if it will really improve the performance. Do you want me to add this?

```diff
diff --git a/actionpack/lib/action_dispatch/middleware/debug_exceptions.rb b/actionpack/lib/action_dispatch/middleware/debug_exceptions.rb
index 2ac604d5e1..7d48eb703b 100644
--- a/actionpack/lib/action_dispatch/middleware/debug_exceptions.rb
+++ b/actionpack/lib/action_dispatch/middleware/debug_exceptions.rb
@@ -121,6 +121,7 @@ def create_template(request, wrapper)
           trace_to_show: wrapper.trace_to_show,
           routes_inspector: routes_inspector(wrapper.exception),
           source_extracts: wrapper.source_extracts,
+          error_highlight_available: wrapper.error_highlight_available?,
           line_number: wrapper.line_number,
           file: wrapper.file
         )
diff --git a/actionpack/lib/action_dispatch/middleware/exception_wrapper.rb b/actionpack/lib/action_dispatch/middleware/exception_wrapper.rb
index 49299ea4d2..d3d725321f 100644
--- a/actionpack/lib/action_dispatch/middleware/exception_wrapper.rb
+++ b/actionpack/lib/action_dispatch/middleware/exception_wrapper.rb
@@ -128,6 +128,14 @@ def source_extracts
       end
     end

+    def error_highlight_available?
+      # ErrorHighlight.spot with backtrace_location keyword is available since error_highlight 0.4.0
+      unless defined?(@@error_highlight_available)
+        @@error_highlight_available = defined?(ErrorHighlight) && Gem::Version.new(ErrorHighlight::VERSION) >= Gem::Version.new("0.4.0")
+      end
+      @@error_highlight_available
+    end
+
     def trace_to_show
       if traces["Application Trace"].empty? && rescue_template != "routing_error"
         "Full Trace"
@@ -174,8 +182,7 @@ def clean_backtrace(*args)
       end

       def extract_source(trace)
-        # ErrorHighlight.spot with backtrace_location keyword is available since error_highlight 0.4.0
-        if defined?(ErrorHighlight) && Gem::Version.new(ErrorHighlight::VERSION) >= Gem::Version.new("0.4.0")
+        if error_highlight_available?
           spot = ErrorHighlight.spot(@exception, backtrace_location: trace)
           if spot
             line = spot[:first_lineno]
diff --git a/actionpack/lib/action_dispatch/middleware/templates/rescues/_source.html.erb b/actionpack/lib/action_dispatch/middleware/templates/rescues/_source.html.erb
index 0ab2967fe4..d0d7a0663e 100644
--- a/actionpack/lib/action_dispatch/middleware/templates/rescues/_source.html.erb
+++ b/actionpack/lib/action_dispatch/middleware/templates/rescues/_source.html.erb
@@ -28,7 +28,7 @@
           </tr>
         </table>
       </div>
-      <%- if defined?(ErrorHighlight) && Gem::Version.new(ErrorHighlight::VERSION) < Gem::Version.new("0.4.0") -%>
+      <%- if defined?(ErrorHighlight) && !error_highlight_available -%>
         <p class="error_highlight_tip">Tip: You may want to add <code>gem 'error_highlight', '&gt;= 0.4.0'</code> into your Gemfile, which will display the fine-grained error location.</p>
       <%- end -%>
     </div>
diff --git a/actionpack/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb b/actionpack/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb
index a68eaab8d7..60654ce3d6 100644
--- a/actionpack/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb
+++ b/actionpack/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb
@@ -11,7 +11,7 @@
   <%= render "rescues/message_and_suggestions", exception: @exception %>
   <%= render "rescues/actions", exception: @exception, request: @request %>

-  <%= render "rescues/source", source_extracts: @source_extracts, show_source_idx: @show_source_idx, error_index: 0 %>
+  <%= render "rescues/source", source_extracts: @source_extracts, error_highlight_available: @error_highlight_available, show_source_idx: @show_source_idx, error_index: 0 %>
   <%= render "rescues/trace", traces: @traces, trace_to_show: @trace_to_show, error_index: 0 %>

   <% if @exception.cause %>
@@ -26,7 +26,7 @@
     </div>

     <div id="<%= wrapper.exception.object_id %>" class="hidden">
-      <%= render "rescues/source", source_extracts: wrapper.source_extracts, show_source_idx: wrapper.source_to_show_id, error_index: index %>
+      <%= render "rescues/source", source_extracts: wrapper.source_extracts, error_highlight_available: wrapper.error_highlight_available, show_source_idx: wrapper.source_to_show_id, error_index: index %>
       <%= render "rescues/trace", traces: wrapper.traces, trace_to_show: wrapper.trace_to_show, error_index: index %>
     </div>
   <% end %>
```

Reply:
> Do you want me to add this?

I think it's better yes.